RavenDB-19524 - 'TombstonesBucketAndEtag' storage index maps revision-tombstones to the wrong bucket

shiranshalom opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19524/TombstonesBucketAndEtag-storage-index-maps-revision-tombstones-to-the-wrong-bucket

### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

shiranshalom committed:
RavenDB-19524 - 'TombstonesBucketAndEtag' storage index maps revision-tombstones to the wrong bucket

karmeli87 commented on src/Raven.Server/Documents/Sharding/ShardedDocumentsStorage.cs:
```diff
@@ -149,7 +149,11 @@ internal static ByteStringContext.Scope GenerateBucketAndEtagIndexKeyForDocument
+        if (tombstoneType == Tombstone.TombstoneType.Revision)
+            return ExtractIdFromKeyAndGenerateBucketAndEtagIndexKey(tx, (int)TombstoneTable.LowerId, etagIndex: (int)TombstoneTable.Etag, ref tvr, out slice);
+
+        return GenerateBucketAndEtagIndexKey(tx, idIndex: (int)DocumentsTable.LowerId, etagIndex: (int)DocumentsTable.Etag, ref tvr, out slice);
```
why `DocumentsTable.LowerId` & `DocumentsTable.Etag`?

shiranshalom replied:
done

karmeli87 commented on src/Raven.Server/Documents/Sharding/ShardedDocumentsStorage.cs:
```diff
@@ -149,7 +149,11 @@ internal static ByteStringContext.Scope GenerateBucketAndEtagIndexKeyForDocument
     {
-        return GenerateBucketAndEtagIndexKey(tx, idIndex: (int)TombstoneTable.LowerId, etagIndex: (int)TombstoneTable.Etag, ref tvr, out slice);
+        var tombstoneType = *(Tombstone.TombstoneType*)tvr.Read((int)TombstoneTable.Type, out int _);
+        if (tombstoneType == Tombstone.TombstoneType.Revision)
```
please check other types as well (but in case there is a problem and complex to fix open a dedicated issue)

shiranshalom replied:
done - we need to do this for all tombstone types :)

shiranshalom committed:
RavenDB-19524 - fixing PR comments#1

shiranshalom committed:
RavenDB-19524 - adding test & implementation for counter tombstones

aviv86 commented on src/Raven.Server/Documents/Sharding/ShardedDocumentsStorage.cs:
```diff
@@ -310,6 +311,11 @@ public ChangeVector GetMergedChangeVectorInBucket(DocumentsOperationContext cont
+        {
+            var counterTombstoneCv = TableValueToChangeVector(context, (int)CounterTombstones.CounterTombstonesTable.ChangeVector, ref result.Result.Reader);
+            merged = merged.MergeWith(counterTombstoneCv, context);
+        }
```
we don't want to take into account the counter tombstones when computing the merged change vector in bucket,
because counter tombstones are not replicated

shiranshalom replied:
done

aviv86 commented on test/SlowTests/Sharding/Issues/RavenDB_19524.cs:
```diff
@@ -0,0 +1,201 @@
+                using (ctx.OpenReadTransaction())
+                {
+                    var tombstones = db.DocumentsStorage.CountersStorage.GetCounterTombstonesFrom(ctx, 0).ToList();
+                    Assert.Equal(1, tombstones.Count);
```
we want to assert the number of counter tombstones in *bucket*, no?
you can use here `ShardedDocumentsStorage.GetItemsByBucket(allocator, counterTombstonesTable, counterTombstonesBucketAndEtagIndex, bucket, etag)`

shiranshalom replied:
done

shiranshalom committed:
RavenDB-19524 - fixing PR comments#2

ppekrol commented:
Please check failing tests

shiranshalom committed:
RavenDB-19524 - fix failing tests

shiranshalom commented:
> Please check failing tests

Done.
I changed a few tests that were affected by this PR. Most of the changes were for `bucketStats.Size` (now we have all revisions/ attachments tombstones in the bucket so the size is bigger). 
For test `BucketDeletionShouldMarkExistingTombstonesAsArtificial` - my expectation here is that after resharding - on Cleanup source (`DeleteBucket`) we will mark as artificial only the tombstones created by deleting the 'live' documents in the bucket, and not the one we deleted intentionally.
